src/frequenz/client/dispatch/test/service.py

```diff
@@ -0,0 +1,264 @@
+                        case "end_criteria":
+                            pb_dispatch.recurrence.end_criteria.CopyFrom(
+                                request.update.recurrence.end_criteria
+                            )
+                        case "freq" | "interval":
+                            setattr(
+                                pb_dispatch.recurrence,
+                                split_path[1],
+                                getattr(request.update.recurrence, split_path[1]),
+                            )
+                        # Fields of type list that need to be copied
+                        case (
+                            "byminutes"
+                            | "byhours"
+                            | "byweekdays"
+                            | "bymonthdays"
+                            | "bymonths"
+                        ):
+                            getattr(pb_dispatch.recurrence, split_path[1])[:] = getattr(
+                                request.update.recurrence, split_path[1]
+                            )[:]
+
+        dispatch = Dispatch.from_protobuf(pb_dispatch)
+        dispatch.update_time = datetime.now(tz=timezone.utc)
+
+        return Empty()
+
+    async def GetMicrogridDispatch(
+        self,
+        request: DispatchGetRequest,
+    ) -> PBDispatch:
+        """Get a single dispatch."""
+        dispatch = next((d for d in self.dispatches if d.id == request.id), None)
+
+        if dispatch is None:
+            error = grpc.RpcError()
+            # pylint: disable=protected-access
+            error._code = grpc.StatusCode.NOT_FOUND  # type: ignore
+            error._details = "Dispatch not found"  # type: ignore
+            # pylint: enable=protected-access
+            raise error
+
+        return dispatch.to_protobuf()
+
+    async def DeleteMicrogridDispatch(
+        self,
+        request: DispatchDeleteRequest,
+    ) -> Empty:
+        """Delete a given dispatch."""
+        num_dispatches = len(self.dispatches)
+        self.dispatches = [d for d in self.dispatches if d.id != request.id]
+
+        if len(self.dispatches) == num_dispatches:
+            error = grpc.RpcError()
+            # pylint: disable=protected-access
+            error._code = grpc.StatusCode.NOT_FOUND  # type: ignore
+            error._details = "Dispatch not found"  # type: ignore
+            # pylint: enable=protected-access
+            raise error
+
+        return Empty()
+
+    # pylint: enable=invalid-name
+
+
+class FakeClient(Client):
+    """Fake client for testing.
+
+    This client uses a fake service to simulate the dispatch api.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the mock client."""
+        super().__init__(grpc.aio.insecure_channel("mock"), "mock")
+        self._stub = FakeService()  # type: ignore
+
+    @property
+    def service(self) -> FakeService:
+        """Get the fake service."""
+        return self._stub  # type: ignore
```

Comment:
I would move this to `client.py` and make `service.py` private (`_service.py`), so you can use it in the tests for the client itself (in this repo) but it is still hidden to users of the client, which shouldn't do anything related to protobuf.

And if possible (and not very annoying for the tests of this repo), I would not make the `service` property public, if you are doing this only to provide access to the `service.dispatches` list, then add a `dispatches` property that returns `self._stub.dispatches`, assuming these dispatches are the wrappers and not the protobuf types.

Reply:
OK, I see now that you are using `__all__` to *hide* the `FakeService`. I'm not sure if this will hide it from the generated docs too (you can try), but still I would recommend going for the splitting approach for 2 reasons:

1. Consistency, other projects are using that approach.
2. The user of this library will import `from ...test.service import FakeClient`. It is a bit weird to have `service` in there.